scripts/parse_vcgs_manifest.py

```diff
@@ -0,0 +1,381 @@
+        path_prefix = os.path.dirname(manifest_path)
+        manifest_filename = os.path.basename(manifest_path)
+        parser = VcgsManifestParser(path_prefix, sample_metadata_project)
+        file_contents = parser.file_contents(manifest_filename)
+        resp = parser.parse_vcgs_manifest(StringIO(file_contents))
+
+        return resp
+
+    def file_path(self, filename):
+        if self.client:
+            return os.path.join('gs://', self.bucket, self.path_prefix, filename)
+        return os.path.join(self.path_prefix, filename)
+
+    def file_contents(self, filename):
+        path = os.path.join(self.path_prefix, filename)
+        if self.client:
+            blob = self.bucket_client.get_blob(path)
+            return blob.download_as_string().decode()
+        else:
+            with open(path) as f:
+                return f.read()
+
+    def file_exists(self, filename):
+        path = os.path.join(self.path_prefix, filename)
+        if self.client:
+            blob = self.bucket_client.get_blob(path)
+            return blob.exists()
+        else:
+            return os.path.exists(path)
+
+    def file_size(self, filename):
+        path = os.path.join(self.path_prefix, filename)
+        if self.client:
+            blob = self.bucket_client.get_blob(path)
+            return blob.size
+        else:
+            return os.path.getsize(path)
+
+    def parse_vcgs_manifest(self, file_pointer):
+        # a sample has many rows
+        sample_map = defaultdict(list)
+
+        reader = csv.DictReader(file_pointer, delimiter='\t')
+        for row in reader:
+            sample_map[row[Columns.SAMPLE_NAME]].append(row)
+
+        # now we can start adding!!
+        sapi = SampleApi()
+        seqapi = SequenceApi()
+
+        # determine if any samples exist
+        external_id_map = sapi.get_sample_id_map_by_external('dev', {'external_ids': list(sample_map.keys())}, allow_missing=True)
+
+        samples_to_update: Dict[Dict] = {}
+        samples_to_add: List[NewSample] = []
+        # by external_sample_id
+        sequencing_to_add: Dict[str, List[NewSequence]] = defaultdict(list)
+        for sample_name in sample_map:
+            logger.info(f'Preparing {sample_name}')
+            rows = sample_map[sample_name]
+            reads = self.parse_reads([r[Columns.FILENAME] for r in rows])
+            # now we have sample / sequencing meta across 4 different rows, so collapse them
+            collapsed_sequencing_meta = {
+                col: ",".join(set(r[col] for r in rows))
+                for col in Columns.sequence_columns()
+            }
+            collapsed_type = self.parse_sequencing_type(
+                sample_name, [r[Columns.LIBRARY_STRATEGY] for r in rows]
+            )
+            collapsed_sample_meta = {
+                col: ",".join(set(r[col] for r in rows))
+                for col in Columns.sample_columns()
+            }
+            collapsed_sample_meta['reads'] = reads
```

Comment:
Should we also add some field for the file type (`fastq`/`cram`/`bam`)? Though probably not a big deal, the loading pipeline would check the extension (or whether the object is a list of lists for fastqs).